GeneratorPgn.py

```diff
@@ -0,0 +1,318 @@
+import chess
+import torch.multiprocessing as mp
+import bz2
+from collections import deque
+import random
+import urllib
+import numpy as np
+
+class GeneratorPgn():
+    def __init__(self, replay_queue: mp.Queue, pgn_path):
+
+        """
+        Iterable class that generates stochastic fen and value samples by parallel running generators.
+        :param replay_queue: target queue for samples
+        :param current_line: currently used line from the data file, representing one game
+        :param board: the chessboard on which the moves are made
+        :param pgn_paths: list of paths to data files
+        :param current_path: index of the path to the currently used data file in the pgn_paths list
+        :param pgn_file: currently used data file
+        """
+
+        self.replay_queue = replay_queue
+        self.current_path = 0
+        self.current_line = []
+        self.board = chess.Board()
+        if type(pgn_path) == str:
+            self.pgn_paths = [pgn_path]
+        else:
+            self.pgn_paths = pgn_path
+        if ".bz2" in self.pgn_paths[0]:
+            self.pgn_file = bz2.open(self.pgn_paths[0])
+        elif "https://" in self.pgn_paths[0]:
+            self.pgn_file = urllib.request.urlopen(self.pgn_paths[0])
+        else:
+            self.pgn_file = open(self.pgn_paths[0])
+
+    def get_line(self):
+        """
+        get a new line with a new game from the currently used data file into the variable current_line.
+        If all lines from the file have already been read, set the next file from the list pgn_paths to the pgn_file and read the first line
+        :return: None
+        """
+        line = self.pgn_file.readline()
+        if type(line) == bytes:
+            line = line.decode("utf-8")
+        if line == "":
+            self.current_path += 1
+            if self.current_path >= len(self.pgn_paths):
+                self.current_path = 0
+            if self.pgn_paths[self.current_path].endswith(".bz2"):

```

Comment:
this is duplicate logic (first occurence: __init__)

Reply:
I don't understand. Could you explain?

Reply:
I mean, that the logic (to support loading normal files, unpacking .bz2 and downloading via http) is in two places. If you'd extract this to a function, you'd have less code repetition (and less places to update somethng if something needs changing)

Reply:
Ah, indeed.